packages/xstate-react/src/useMachine.ts

```diff
@@ -163,44 +166,47 @@ export function useMachine<
   >([]);
 
   useLayoutEffect(() => {
-    service.onTransition((currentState) => {
-      if (currentState.changed) {
-        setState(currentState);
-      }
-
-      if (currentState.actions.length) {
-        const reactEffectActions = currentState.actions.filter(
-          (action): action is ReactActionObject<TContext, TEvent> => {
-            return (
-              typeof action.exec === 'function' &&
-              '__effect' in (action as ReactActionObject<TContext, TEvent>).exec
-            );
-          }
-        );
-
-        const [effectActions, layoutEffectActions] = partition(
-          reactEffectActions,
-          (action): action is ReactActionObject<TContext, TEvent> => {
-            return action.exec.__effect === ReactEffectType.Effect;
-          }
-        );
-
-        effectActionsRef.current.push(
-          ...effectActions.map<ActionStateTuple<TContext, TEvent>>(
-            (effectAction) => [effectAction, currentState]
-          )
-        );
-
-        layoutEffectActionsRef.current.push(
-          ...layoutEffectActions.map<ActionStateTuple<TContext, TEvent>>(
-            (layoutEffectAction) => [layoutEffectAction, currentState]
-          )
-        );
-      }
-    });
-
-    // if service.state has not changed React should just bail out from this update
-    setState(service.state);
+    service
+      .onTransition((currentState) => {
+        // Only change the current state if:
+        // - the incoming state is the initial state (since it might have new actors)
+        // - OR the incoming state actually changed
+        if (currentState.changed || currentState.changed === undefined) {
+          setState(currentState);
+        }
+
+        if (currentState.actions.length) {
+          const reactEffectActions = currentState.actions.filter(
+            (action): action is ReactActionObject<TContext, TEvent> => {
+              return (
+                typeof action.exec === 'function' &&
+                '__effect' in
+                  (action as ReactActionObject<TContext, TEvent>).exec
+              );
+            }
+          );
+
+          const [effectActions, layoutEffectActions] = partition(
```

Comment:
could you add a test that those initial "react" actions are executed in the very first commit phase of a component? from what I see it works like this (which is IMHO good), but this also seems like a potentially fragile logic, so would be great to have a test for this

Reply:
What would a test for this look like? Might be a bit too implementation-detail oriented. The `'layout effects should happen after normal actions'` test covers the order, ensuring that a `layout effect` happens after a `non-effect` and a normal `effect`.

Reply:
> The 'layout effects should happen after normal actions' test covers the order, ensuring that a layout effect happens after a non-effect and a normal effect.

Oh, forgot about this test as it isn't part of this PR specifically. It is focused on order between normal and effect actions though - we only know from it that effect actions execute **after** regular actions, but we don't quite ensure when they execute.

You could extend that existing test or write a new one like this:

```js
  it('initial effect actions should execute during the very first commit phase', (done) => {
    let commitPhaseCounter = 0;

    const machine = createMachine(
      {
        initial: 'active',
        states: {
          active: {
            entry: [
              asLayoutEffect(() => {
                expect(commitPhaseCounter).toBe(1);
              }),
              asEffect(() => {
                expect(commitPhaseCounter).toBe(1);
              }),
            ]
          }
        }
      },
    );

    const App = () => {
      React.useLayoutEffect(() => {
        commitPhaseCounter++
      })
      useMachine(machine);

      return <div />;
    };

    render(<App />);
    done();
  });
```

Reply:
Added that test 👍